### PR TITLE
chore: RFC #2445 - 2020-04-23 - Log Schema Facilities & Standards

### LINKED ISSUE #2445: Review all sinks to ensure tls settings are set correctly
Looks like with a lot of the movement that we've had with tls types and http clients over the past month or so has caused some confusion around when tls settings are enabled or not by default. We should do a re-review of our sinks and if some sink should have tls settings enabled by default.

### PR DESCRIPTION
**Pull request by binarylogic**
Closes #2445

[Rendered](https://github.com/timberio/vector/blob/schema-rfc/rfcs/2020-04-23-2445-log-schema.md)

Ref #2272, #2273, #2147, #2269, #2270, #2271, #1150, #2477, #2480

### CODE REVIEW

**Comment by binarylogic** [comment]
@Hoverbear @lukesteensen this is ready for review. I'd like to get @ktff on this once he finishes his current work.

File: rfcs/2020-04-23-2445-log-schema.md [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+   schemas are arguably the biggest source of lock-in when coming from other
+   tools.
+4. We can be flexible and adopt new standards as they emerge. For example,
+   OpenTelemetry is [working on a schema standard][ot_log_data_model] that I
+   would like Vector to _optionally_ adopt when ready.
+
+## Drawbacks
+
+1. It's more complicated.
+2. I'm unsure if this will affect performance negatively. Adding more fields
```
Comment (ktff) [inline rfcs/2020-04-23-2445-log-schema.md:94]: It shouldn't affect performance any more than the alternatives. And with the ability to turnoff the fields, users could use it to trade extra fields for extra performance. Some of the presets could also, by default, turn off some of them. Example: `slim` preset could have only the most important fields set, and others disabled.

File: rfcs/2020-04-23-2445-log-schema.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
```
Comment (Hoverbear) [inline rfcs/2020-04-23-2445-log-schema.md:94]: If we are adding fields statically (eg we are adding fields to a struct, not string keys to a hashmap) we shouldn't see a huge performance cost except for serializing/deserializing.

File: rfcs/2020-04-23-2445-log-schema.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+1. Move all component level keys to the global `log_schema` set of options.
+2. Remove all component level `*_key` options in favor of the new `log_schema.*`
+   options.
+3. Define a namespace for metadata, such as `log_schema.meta_key`. This is
+   intended for internal Vector metadata that does not describe the log
+   directly.
+4. If the field does not directly describe the log, or it's origine, then it
+   should go under the new `log_schema.meta_key`.
+5. Setting any of these fields to `""` disables the field.
```
Comment (Hoverbear) [inline rfcs/2020-04-23-2445-log-schema.md:52]: Can we note that internally Vector *must* store these as `Option<_>` and not store full empty `String` types.

File: rfcs/2020-04-23-2445-log-schema.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+  - [ ] Do _not_ override the log field if it already exists.
+  - [ ] For any non-leaf keys, if the value is set to `""` then it should insert all descendant keys at the root of the log record.
+  - [ ] For any leaf keys, if the value is set to `""` then it should skip inserting the value. This effectively disables the key.
+
+#### Sources
+
+- [ ] `docker` source
+  - [ ] Set the `log_schema.meta.source_type_key` to `docker`.
+  - [ ] Set the`log_schema.containers_key` option.
+  - [ ] Set the`log_schema.containers.created_at_key` option to the previous `contrainer_created_at` field.
```
Comment (Hoverbear) [inline rfcs/2020-04-23-2445-log-schema.md:165]:
```suggestion
  - [ ] Set the`log_schema.containers.created_at_key` option to the previous `container_created_at` field.
```

File: rfcs/2020-04-23-2445-log-schema.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+1. Move all component level keys to the global `log_schema` set of options.
+2. Remove all component level `*_key` options in favor of the new `log_schema.*`
+   options.
+3. Define a namespace for metadata, such as `log_schema.meta_key`. This is
+   intended for internal Vector metadata that does not describe the log
+   directly.
+4. If the field does not directly describe the log, or it's origine, then it
+   should go under the new `log_schema.meta_key`.
+5. Setting any of these fields to `""` disables the field.
+6. Offer a "legacy" setting that sets all of the fields below to their current
```
Comment (Hoverbear) [inline rfcs/2020-04-23-2445-log-schema.md:53]: Can we set a timeline on deprecating this? Like 2.0 or something?

File: rfcs/2020-04-23-2445-log-schema.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+  - Keep the `log_schema.message_key` option (default `message`).
+  - [ ] Add a new `log_schema.stream_key` option (default `stream`).
+  - Keep the `log_schema.timestamp_key` option (default `timestamp`).
+- [ ] Add a new `log_schema.containers_key` option (default `containers`).
+  - [ ] Add a new `log_schema.containers.created_at_key` option (default `created_at`).
+  - [ ] Add a new `log_schema.containers.id_key` option (default `id`).
+  - [ ] Add a new `log_schema.containers.name_key` option (default `name`).
+  - [ ] Add a new `log_schema.containers.image_key` option (default `image`).
+  - [ ] Add a new `log_schema.containers.labels_key` option (default `labels`).
+- [ ] Add a new `log_schema.file_key` option (default `file`).
```
Comment (Hoverbear) [inline rfcs/2020-04-23-2445-log-schema.md:128]: So what if a user wants to have two `file` sources with different `file` keys? These options are global no?

File: rfcs/2020-04-23-2445-log-schema.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+  - [ ] Add a new `log_schema.file.path_key` option (default `path`).
```
Comment (Hoverbear) [inline rfcs/2020-04-23-2445-log-schema.md:129]: This doesn't seem to namespace sources/sinks/transforms and may lead to awkward collisions later -- as well as a painful migration. Suggest namespacing fully.

**Comment by Hoverbear** [comment]
This is a great idea, will be super excited to see this happen! :)

**Comment by binarylogic** [comment]
Closing this based on external discussions. Vector plans to take an unopinionated stance on schemas. We plan to use internal metadata to describe fields in order to map them appropriately in sources and sinks. We'll follow up with a separate RFC that outlines this approach.

**Pull request closed without merging**